module/executiondatasync/optimistic_syncing/core_impl_test.go

```diff
@@ -0,0 +1,360 @@
+
+		err := core.Download(ctx)
+
+		assert.Error(t, err)
+		assert.ErrorIs(t, err, assert.AnError)
+		assert.Nil(t, core.executionData)
+
+		txResultErrMsgsRequester.AssertNotCalled(t, "Request")
+	})
+
+	t.Run("transaction result error messages request error", func(t *testing.T) {
+		core, execDataRequester, txResultErrMsgsRequester := createTestCoreImpl(t)
+		ctx := context.Background()
+
+		expectedExecutionData := unittest.BlockExecutionDataFixture()
+		execDataRequester.On("RequestExecutionData", ctx).Return(expectedExecutionData, nil)
+
+		txResultErrMsgsRequester.On("Request", ctx).Return(([]flow.TransactionResultErrorMessage)(nil), assert.AnError)
+
+		err := core.Download(ctx)
+
+		assert.Error(t, err)
+		assert.ErrorIs(t, err, assert.AnError)
+		assert.NotNil(t, core.executionData)
+	})
+}
+
+// TestCoreImpl_Index tests the Index method which processes downloaded data using the real InMemoryIndexer.
+// The Index method must be called after Download has successfully retrieved execution data.
+// Calling Index without prior successful Download will result in a panic due to nil execution data.
+func TestCoreImpl_Index(t *testing.T) {
+	t.Run("successful indexing", func(t *testing.T) {
+		core, _, _ := createTestCoreImpl(t)
+		ctx := context.Background()
+
+		// Create execution data with the SAME block ID as the execution result
+		executionData := unittest.BlockExecutionDataFixture(
+			unittest.WithBlockExecutionDataBlockID(core.executionResult.BlockID),
+		)
+		core.executionData = execution_data.NewBlockExecutionDataEntity(
+			core.executionResult.ExecutionDataID,
+			executionData,
+		)
+
+		core.txResultErrMsgsData = []flow.TransactionResultErrorMessage{
+			{
+				TransactionID: unittest.IdentifierFixture(),
+				Index:         0,
+				ErrorMessage:  "test error",
+				ExecutorID:    unittest.IdentifierFixture(),
+			},
+		}
+
+		err := core.Index(ctx)
+		require.NoError(t, err)
+	})
+
+	t.Run("block ID mismatch", func(t *testing.T) {
+		core, _, _ := createTestCoreImpl(t)
+		ctx := context.Background()
+
+		// Create execution data with a DIFFERENT block ID than expected
+		executionData := unittest.BlockExecutionDataFixture()
+		core.executionData = execution_data.NewBlockExecutionDataEntity(
+			unittest.IdentifierFixture(),
+			executionData,
+		)
+
+		err := core.Index(ctx)
+
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "invalid block execution data")
+		assert.Contains(t, err.Error(), "expected block_id")
+	})
+}
+
+// TestCoreImpl_Persist tests the Persist method which saves indexed data using the real Persister.
+func TestCoreImpl_Persist(t *testing.T) {
+	t.Run("successful persistence", func(t *testing.T) {
+		// Create mocks with proper expectations
+		mockDB := storagemock.NewDB(t)
+		mockBatch := storagemock.NewBatch(t)
+
+		// Set up successful batch operations
+		mockBatch.On("Commit").Return(nil)
+		mockBatch.On("Close").Return(nil).Maybe()
```

Comment:
close must be called if the batch is created, otherwise there's a leak
```suggestion
		mockBatch.On("Close").Return(nil)
```